tidy3d/components/data/monitor_data.py

```diff
@@ -424,6 +424,194 @@ class ModeSolverData(ModeSolverDataset, ElectromagneticFieldData):
+                )
+                # Transform "sorting w.r.t. neighbor" to "sorting w.r.t. to f0_ind"
+                sorting[freq_id, :] = sorting_one_mode[sorting[freq_id - step, :]]
+                overlap[freq_id, :] = overlap_one_mode[sorting[freq_id - step, :]]
+                # Check for discontinuities and show warning if any
+                for mode_ind in list(np.nonzero(overlap[freq_id, :] < overlap_thresh)[0]):
+                    log.warning(
+                        f"WARNING: Mode '{mode_ind}' appears to undergo a discontinuous change "
+                        f"between frequencies '{self.monitor.freqs[freq_id]}' "
+                        f"and '{self.monitor.freqs[freq_id - step]}' "
+                        f"(overlap: '{overlap[freq_id, mode_ind]:.2f}')."
+                    )
+                # Reassign for the next iteration
+                data_template = data_to_sort
+                norm_template = norm_to_sort
+
+        # Rearrange modes using computed sorting values
+        mode_data_sorted = self._reorder_modes(
+            sorting=sorting,
+            gauge_indices=gauge_indices,
+            track=track,
+        )
+
+        return mode_data_sorted
+
+    # pylint:disable=too-many-arguments
+    def _find_ordering_one_freq(
+        self,
+        data_template: ModeSolverData,
+        norm_template: Numpy,
+        data_to_sort: ModeSolverData,
+        norm_to_sort: Numpy,
+        overlap_thresh: float = 0.75,
+    ) -> Tuple[Numpy, Numpy]:
+        """Find new ordering of modes in data_to_sort based on their similarity to data_template."""
+
+        num_modes = data_template.n_complex.sizes["mode_index"]
+
+        # Check whether modes already match
+        current_overlaps = np.abs(data_template.dot(data_to_sort))
+        current_overlaps = current_overlaps.data.ravel() / norm_template / norm_to_sort
+
+        if np.all(current_overlaps > overlap_thresh):
+            return np.arange(num_modes), current_overlaps
+
+        # Compute an overlap matrix, dot products of each template mode with each mode to sort
+        overlaps = np.zeros((num_modes, num_modes))
+        for mode_index in range(num_modes):
+            # Get one mode from data_to_sort
+            one_mode = data_to_sort.sel_mode_index(mode_index)
+            dot_product_raw = np.abs(data_template.dot(one_mode).data.ravel())
+            # Appropriate normalization
+            overlaps[:, mode_index] = dot_product_raw / norm_template / norm_to_sort[mode_index]
+        # Find the most similar modes and corresponding overlap values
+        closest_pairs, closest_overlaps = ModeSolverData._find_closest_pairs(overlaps)
+        return closest_pairs, closest_overlaps
+
+    def _find_closest_pairs(arr: Numpy) -> Tuple[Numpy, Numpy]:
+        """Given an overlap matrix pair row and column entries."""
+
+        n, k = np.shape(arr)
+        assert n == k
```

Comment:
want to avoid assert statements unless in tests, let's raise an error here if they are not equal instead.